src/core/version_neg.c

```diff
@@ -0,0 +1,384 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    This file defines the logic for version negotiation.
+
+--*/
+
+#include "precomp.h"
+#ifdef QUIC_CLOG
+#include "version_neg.c.clog.h"
+#endif
+
+typedef struct QUIC_COMPATIBLE_VERSION_MAP {
+    const uint32_t OriginalVersion;
+    const uint32_t CompatibleVersion;
+} QUIC_COMPATIBLE_VERSION_MAP;
+
+const QUIC_COMPATIBLE_VERSION_MAP CompatibleVersionsMap[] = {
+    {QUIC_VERSION_DRAFT_29, QUIC_VERSION_DRAFT_29},
+    {QUIC_VERSION_MS_1, QUIC_VERSION_MS_1},
+    {QUIC_VERSION_MS_1, QUIC_VERSION_1},
+    {QUIC_VERSION_1, QUIC_VERSION_1},
+    {QUIC_VERSION_1, QUIC_VERSION_MS_1}
+};
+
+//
+// This list is the versions the server advertises support for.
+//
+const uint32_t DefaultSupportedVersionsList[] = {
+    QUIC_VERSION_1,
+    QUIC_VERSION_MS_1,
+    QUIC_VERSION_DRAFT_29
+};
+const uint32_t DefaultSupportedVersionsListLength = ARRAYSIZE(DefaultSupportedVersionsList);
+
+const uint32_t DefaultCompatibleVersionsList[] = {
+    QUIC_VERSION_1,
+    QUIC_VERSION_MS_1
+};
+const uint32_t DefaultCompatibleVersionsListLength = ARRAYSIZE(DefaultCompatibleVersionsList);
```

Comment:
I feel like we should always calculate this, so we don't have to depend on the developer correctly updating it as the default supported list and compatibility list change over time.

Reply:
Should it live on the MsQuicLib object for the default list? The benefit here is that it saves memory and computation time for the vast majority of apps. (and it's read-only memory)

Reply:
Feel free to save it where ever you think best. Calculating it once for default apps at library start up is ~free.